### ISSUE TITLE
Python port?

### ISSUE THREAD

**Issue by krober10nd** [dengwirda/inpoly#1]
Hey Darren, 

Hope all is well. I'm thinking about porting this over to pure Python with numpy. Have you done this already? Just thought I'd ask before embarking on a new weekend project.

**Comment by krober10nd** [comment]
I could just pybind11 your CPP code actually so you can call it with numpy arrays

**Comment by dengwirda** [comment]
Hi Keith,

I'm not sure how the `inpoly` algorithm would work out in pure `python` --- the core is definitely loop-based, which I suspect may be quite slow in `numpy` / `python`.

The approach I took with `octave` was to write the [backend part in `c++`](https://github.com/dengwirda/inpoly/blob/master/inpoly2_oct.cpp) with the [frontend *.m part](https://github.com/dengwirda/inpoly/blob/master/inpoly2.m) handling the vector array operations --- possibly something like this could work with `numpy` too.

**Comment by krober10nd** [comment]
Hey Darren, 

Okay... well I took a stab at it anyway and I think it came out well. I get the same answer as your version in MATLAB. See below...

> I'm not sure how the inpoly algorithm would work out in pure python --- the core is definitely loop-based, which I suspect may be quite slow in numpy / python.

It's worth noting that Python has no dedicated `inpolygon` function. Matplotlib as has something called a `Path` which has a method `contains_points` but that only works for one `Path` at one time and becomes horribly slow.  Anyway, I didn't do any profiling effort on this but `numba` could easily speed this up quite a bit by adding some simple decorators. 

Do you mind if I go ahead and make a pypi package with it (after some more testing)? It goes without saying, I'll put all the credit to you and carry over your license file in the pip package. I think this would increase the visibility of your good work.

Here's the implementation port (your demo test and the random point set is attached). 

```
 import numpy as np
 
 
 def inpoly(vert, node, edge=None, ftol=4.9485e-14):
     """A port of Darren Engwirda's `inpoly` routine into Python.
 
        Returns the "inside/outside" status for a set of vertices VERT and
        a polygon {NODE,EDGE}  embedded in a two-dimensional plane. General
        non-convex and multiply-connected polygonal regions can be handled.
 
 
     Parameters
     ----------
     vert: array-like
         VERT is an N-by-2 array of XY coordinates to query.
 
     node: array-like
         NODE is an M-by-2 array of polygon vertices.
 
     edge: array-like, optional
         EDGE is a P-by-2 array of edge indexing. Each
         row in EDGE represents an edge of the polygon, such that
         node(edge(KK,1),:) and node(edge(KK,2),:) are the coord-
         inates of the endpoints of the KK-TH edge. If the argum-
         ent EDGE is omitted it assumed that the vertices in NODE
         are connected in ascending order.
 
     ftol: float, otpional
         FTOL is a floating-point tolerance for boundary comparisons.
         By default, FTOL = EPS ^ 0.85.
 
 
     Returns:
     --------
     STAT: array-like
         STAT is an associated N-by-1 logical array,
         with STAT(II) = TRUE if VERT(II,:) is an interior point.
 
     BNDS: array-like
         N-by-1 logical array BNDS, with BNDS(II) = TRUE if VERT(II,:)
         lies "on" a boundary segment,
 
 
     Notes
     -----
 
     ALL CREDIT GOES TO THE ORIGINAL AUTHOR: Darren Engwirda
 
     Darren Engwirda : 2017 --
     Email           : de2363@columbia.edu
     Last updated    : 31/03/2020
 
 
     Author of port:
 
     Keith J. Roberts (2020)
     Email: krober@usp.br
     Date: 2020-13-09
 
     """
 
     nnod = len(node)
     nvrt = len(vert)
 
     # create edge if not supplied
     if edge is None:
         edge = np.vstack((np.arange(0, nnod - 1), np.arange(1, nnod))).T
         edge = np.concatenate((edge, [[nnod - 1, 0]]))
 
     STAT = np.zeros((nvrt))
     BNDS = np.zeros((nvrt))
 
     # prune points using bbox
     mask = (
         (vert[:, 0] >= np.min(node[:, 0]))
         & (vert[:, 0] <= np.max(node[:, 0]))
         & (vert[:, 1] >= np.min(node[:, 1]))
         & (vert[:, 1] <= np.max(node[:, 1]))
     )
 
     vert = vert[mask]
 
     # flip to ensure y-axis is the `long` axis
     vmin = np.amin(vert, axis=0)
     vmax = np.amax(vert, axis=0)
     ddxy = vmax - vmin
 
     lbar = np.sum(ddxy) / 2.0
 
     if ddxy[0] > ddxy[1]:
         vert = vert[:, [1, 0]]
         node = node[:, [1, 0]]
 
     # sort points via y-value
     swap = node[edge[:, 1], 1] < node[edge[:, 0], 1]
 
     tmp = edge[swap]
     edge[swap, :] = tmp[:, [1, 0]]
 
     ivec = np.argsort(vert[:, 1])
     vert = vert[ivec]
 
     t_stat, t_bnds = _inpoly(vert, node, edge, ftol, lbar)
 
     stat = np.ones(len(vert))
     bnds = np.ones(len(vert))
     stat[ivec] = t_stat
     bnds[ivec] = t_bnds
 
     STAT[mask] = stat
     BNDS[mask] = bnds
 
     return STAT, BNDS
 
 
 def _inpoly(vert, node, edge, ftol, lbar):
 
     feps = ftol * lbar ** 2
     veps = ftol * lbar ** 1
 
     _nvrt = len(vert)
     _nedge = len(edge)
 
     _stat = np.zeros((_nvrt))
     _bnds = np.zeros((_nvrt))
 
     # loop over polygon edges
     for epos in range(_nedge):
 
         inod = edge[epos, 0]
         jnod = edge[epos, 1]
 
         # calc. edge bounding-box
         yone = node[inod, 1]
         ytwo = node[jnod, 1]
         xone = node[inod, 0]
         xtwo = node[jnod, 0]
 
         xmin = np.minimum(xone, xtwo)
         xmax = np.maximum(xone, xtwo)
 
         xmax += veps
 
         ymin = yone - veps
         ymax = ytwo + veps
 
         ydel = ytwo - yone
         xdel = xtwo - xone
 
         # find top VERT[:,1]<YONE
         ilow = 0  # for python
         iupp = _nvrt
 
         while ilow < iupp - 1:  # binary search
             imid = int(ilow + np.floor((iupp - ilow) / 2))
             if vert[imid, 1] < ymin:
                 ilow = imid
             else:
                 iupp = imid
 
         if vert[ilow, 1] >= ymin:
             ilow -= 1
 
         # calc. edge-intersection
         for jpos in range(ilow + 1, _nvrt):
 
             if _bnds[jpos]:
                 continue
 
             xpos = vert[jpos, 0]
             ypos = vert[jpos, 1]
 
             if ypos <= ymax:
                 if xpos >= xmin:
                     if xpos <= xmax:
                         # compute crossing number
                         mul1 = ydel * (xpos - xone)
                         mul2 = xdel * (ypos - yone)
                         if feps >= np.abs(mul2 - mul1):
                             # BNDS -- approx. on edge
                             _bnds[jpos] = 1
                             _stat[jpos] = 1
                         elif (ypos == yone) & (xpos == xone):
                             # BNDS -- match about ONE
                             _bnds[jpos] = 1
                             _stat[jpos] = 1
                         elif (ypos == ytwo) & (xpos == xtwo):
                             # BNDS -- match about TWO
                             _bnds[jpos] = 1
                             _stat[jpos] = 1
                         elif mul1 < mul2:
                             if (ypos >= yone) & (ypos < ytwo):
                                 # advance crossing number
                                 _stat[jpos] = 1 - _stat[jpos]
 
                 else:
                     if (ypos >= yone) & (ypos < ytwo):
                         # advance crossing number
                         _stat[jpos] = 1 - _stat[jpos]
             else:
                 # done -- due to the sort
                 break
     return _stat, _bnds
```                                           

Here's the test to demonstrate it apparently works
```
 import numpy as np
 import matplotlib.pyplot as plt
 
 from inpoly import inpoly
 
 node = np.array([[4, 0], [8, 4], [4, 8], [0, 4], [3, 3], [5, 3], [5, 5], [3, 5]])
 
 edge = np.array([[1, 2], [2, 3], [3, 4], [4, 1], [5, 6], [6, 7], [7, 8], [8, 5]])
 
 edge -= 1  # for Python
 
 points = np.loadtxt("test_points.txt", delimiter=",")
 
 inside, on = inpoly(points, node, edge)
 
 fig, ax = plt.subplots()
 plt.plot(points[inside == 1, 0], points[inside == 1, 1], "b.")
 
 plt.plot(node[:4, 0], node[:4, 1], "r-")
 plt.plot(node[4:, 0], node[4:, 1], "g-")
 
 plt.show()
```
![Inpoly_test](https://user-images.githubusercontent.com/18619644/93032269-0bb77d80-f607-11ea-9615-d334bc4ec561.png)
[test_points.txt](https://github.com/dengwirda/inpoly/files/5215452/test_points.txt)

**Comment by dengwirda** [comment]
Okay, looks pretty good @krober10nd! I'm happy for this to go onto `pypi` at some point, but it'd be interesting to do a little more performance testing first I think. 

Sometimes it seems the loop overhead for `python` can be really awful, so it's probably worthwhile doing a few `python` vs `MATLAB` comparisons on large problems and making sure `inpoly.py` is indeed running fast. These days `MATLAB` can apparently `JIT`-compile loops to be very speedy, but I'm not sure `python` has this option. Definitely my experience with `Octave` is that large loop overhead can completely ruin performance, and that a `c++` backend can be useful in such cases...

I need to get a few other things done over the coming days, but can potentially look more at this at the end of the week?

**Comment by krober10nd** [comment]
> I'm happy for this to go onto pypi at some point, but it'd be interesting to do a little more performance testing first I think.

Sure thing. I'm interesting in it as well. I heavily use this predicate.

> Sometimes it seems the loop overhead for python can be really awful, so it's probably worthwhile doing a few python vs MATLAB comparisons on large problems and making sure inpoly.py is indeed running fast. These days MATLAB can apparently JIT-compile loops to be very speedy, but I'm not sure python has this option. Definitely my experience with Octave is that large loop overhead can completely ruin performance, and that a c++ backend can be useful in such cases...

Yes, I've had similar performance experiences. I normally try to avoid Python for this reason, but my MATLAB license days are over now. I could port this Python into C++ with just the standard library and then use pybind11 to wrap Python around it (like you did for Octave). I have some experience with packaging Python for this approach.

Let's see how the pure Python performs (I expect poorly too) but for everyday stuff could be a useful package to have lying around.

**Comment by krober10nd** [comment]
Hey, I starting testing with some "real" data. With a multi-part polygon with 3,296 edges and 1,382,395 query points, it took 11.359922170639038 seconds (pure Python)

With replacing the Python's JIT with numba's (just around the kernel `_inpoly`) it reduced the time to 0.9602460861206055 seconds.

```
@jit(nopython=True)
 def _inpoly(vert, node, edge, ftol, lbar):
```

Also, if you have a NaN-delimited polygon, have to make sure to do this at the masking stage....

```
     # prune points using bbox
     mask = (
         (vert[:, 0] >= np.nanmin(node[:, 0]))
         & (vert[:, 0] <= np.nanmax(node[:, 0]))
         & (vert[:, 1] >= np.nanmin(node[:, 1]))
         & (vert[:, 1] <= np.nanmax(node[:, 1]))
     )
```

**Comment by dengwirda** [comment]
@krober10nd, I've made an `inpoly-python` repo that we can work with on this --- I think I've managed to clean-up the pure-Python version a little, and it seems to run around twice as fast for me (on some larger problems).

It'd be great indeed if `numba` and better `jit`-ing became available in Python in general, though I'm not sure it's quite there yet. In fact, `numba` crashes for me on `ubuntu` when `pip3 install`-ing...

Anyway, `inpoly-python` is still a work in progress, though there are a few examples there now to look at --- let me know how this looks to you.
